File: src/sc_neurocore/neurons/models/dcn_neuron.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Deep Cerebellar Nuclei (DCN) Neuron

from __future__ import annotations

import math
from dataclasses import dataclass, field


def _safe_rate(a: float, vhalf: float, v: float, k: float, fallback: float) -> float:
    d = v + vhalf
    if abs(d) < 1e-7:
        return fallback
    return a * d / (1.0 - math.exp(-d / k))


@dataclass
class DCNNeuron:
    """Deep cerebellar nuclei neuron — main output of the cerebellum.

    WB Na⁺/K⁺ core + T-type Ca²⁺ (rebound bursting), Ih (pacemaker),
    persistent Na⁺ (subthreshold), Ca²⁺-dependent AHP. 7 currents total.

    Reference: Llinás & Mühlethaler (1988) J Physiol 404:241;
    Jahnsen (1986) J Physiol 372:129.
    """

    v: float = -60.0
    h: float = 0.6
    n: float = 0.32
    p: float = 0.01
    s: float = 0.8
    r: float = 0.1
    ca: float = 0.05
    g_na: float = 35.0
    g_nap: float = 0.5
    g_k: float = 9.0
    g_t: float = 0.1
    g_ahp: float = 2.0
    g_h: float = 0.02
    g_l: float = 0.2
    e_na: float = 55.0
    e_k: float = -90.0
    e_ca: float = 120.0
    e_h: float = -40.0
    e_l: float = -65.0
    c_m: float = 1.0
    phi: float = 5.0
    tau_ca: float = 150.0
    kd_ahp: float = 0.5
    dt: float = 0.5
    v_threshold: float = -20.0
    gain: float = 1.0
    _sub_steps: int = field(default=20, repr=False)

    def __post_init__(self) -> None:
        self._validate_state()

    def step(self, current: float = 0.0) -> int:
        self._validate_state()
        if not math.isfinite(current):
            raise ValueError("current must be finite")

        inp = self.gain * current
        sub_dt = self.dt / self._sub_steps
        fired = 0
        v = self.v
        h = self.h
        n = self.n
        p = self.p
        s = self.s
        r = self.r
        ca = self.ca

        for _ in range(self._sub_steps):
            alpha_m = _safe_rate(0.1, 35.0, v, 10.0, 1.0)
            beta_m = 4.0 * math.exp(-(v + 60.0) / 18.0)
            m_inf = alpha_m / (alpha_m + beta_m)
            alpha_h = 0.07 * math.exp(-(v + 58.0) / 20.0)
            beta_h = 1.0 / (1.0 + math.exp(-(v + 28.0) / 10.0))
            alpha_n = _safe_rate(0.01, 34.0, v, 10.0, 0.1)
            beta_n = 0.125 * math.exp(-(v + 44.0) / 80.0)

            p_inf = 1.0 / (1.0 + math.exp(-(v + 48.0) / 5.0))
            tau_p = 5.0 + 15.0 / max(0.01, 1.0 + ((v + 48.0) / 10.0) ** 2)

            m_t_inf = 1.0 / (1.0 + math.exp(-(v + 52.0) / 5.0))
            s_inf = 1.0 / (1.0 + math.exp((v + 60.0) / 6.5))
            tau_s = 20.0 + 50.0 / (1.0 + math.exp((v + 65.0) / 10.0))

            r_inf = 1.0 / (1.0 + math.exp((v + 80.0) / 10.0))
            tau_r = 100.0 + 200.0 / (1.0 + math.exp((v + 70.0) / 10.0))

            h += sub_dt * self.phi * (alpha_h * (1.0 - h) - beta_h * h)
            n += sub_dt * self.phi * (alpha_n * (1.0 - n) - beta_n * n)
            p += sub_dt * (p_inf - p) / tau_p
            s += sub_dt * (s_inf - s) / tau_s
            r += sub_dt * (r_inf - r) / tau_r

            i_t = self.g_t * m_t_inf**2 * s * (v - self.e_ca)
            ca_entry = -i_t * 0.001 if i_t < 0.0 else 0.0
            ca += sub_dt * (ca_entry - ca / self.tau_ca)
            ca = max(0.0, ca)

            ahp_inf = ca**2 / (ca**2 + self.kd_ahp**2)

            i_na = self.g_na * m_inf**3 * h * (v - self.e_na)
            i_nap = self.g_nap * p * (v - self.e_na)
            i_k = self.g_k * n**4 * (v - self.e_k)
            i_ahp = self.g_ahp * ahp_inf * (v - self.e_k)
            i_h = self.g_h * r * (v - self.e_h)
            i_l = self.g_l * (v - self.e_l)

            dv_val = (-i_na - i_nap - i_k - i_t - i_ahp - i_h - i_l + inp) / self.c_m
            v += sub_dt * dv_val

            if v >= self.v_threshold:
                fired = 1
                v = -60.0
                s *= 0.5
                ca += 0.5

        candidates = (v, h, n, p, s, r, ca)
        if not all(math.isfinite(value) for value in candidates):
            raise ValueError("DCN candidate state must be finite")

        self.v = max(-100.0, min(60.0, v))
        self.h = max(0.0, min(1.0, h))
        self.n = max(0.0, min(1.0, n))
        self.p = max(0.0, min(1.0, p))
        self.s = max(0.0, min(1.0, s))
        self.r = max(0.0, min(1.0, r))
        self.ca = max(0.0, ca)
        return fired

    def reset(self) -> None:
        self.v = -60.0
        self.h = 0.6
        self.n = 0.32
        self.p = 0.01
        self.s = 0.8
        self.r = 0.1
        self.ca = 0.05
        self._validate_state()

    def _validate_state(self) -> None:
        values = (
            self.v,
            self.h,
            self.n,
            self.p,
            self.s,
            self.r,
            self.ca,
            self.g_na,
            self.g_nap,
            self.g_k,
            self.g_t,
            self.g_ahp,
            self.g_h,
            self.g_l,
            self.e_na,
            self.e_k,
            self.e_ca,
            self.e_h,
            self.e_l,
            self.c_m,
            self.phi,
            self.tau_ca,
            self.kd_ahp,
            self.dt,
            self.v_threshold,
            self.gain,
        )
        if not all(math.isfinite(value) for value in values):
            raise ValueError("DCN state and parameters must be finite")
        for name in ("h", "n", "p", "s", "r"):
            value = getattr(self, name)
            if not 0.0 <= value <= 1.0:
                raise ValueError(f"{name} gate must be in [0, 1]")
        if self.ca < 0.0:
            raise ValueError("ca must be non-negative")
        if any(
            value < 0.0
            for value in (
                self.g_na,
                self.g_nap,
                self.g_k,
                self.g_t,
                self.g_ahp,
                self.g_h,
                self.g_l,
            )
        ):
            raise ValueError("conductances must be non-negative")
        if self.c_m <= 0.0:
            raise ValueError("c_m must be positive")
        if self.phi <= 0.0:
            raise ValueError("phi must be positive")
        if self.tau_ca <= 0.0:
            raise ValueError("tau_ca must be positive")
        if self.kd_ahp <= 0.0:
            raise ValueError("kd_ahp must be positive")
        if self.dt <= 0.0:
            raise ValueError("dt must be positive")
        if self.gain < 0.0:
            raise ValueError("gain must be non-negative")
        if self._sub_steps <= 0:
            raise ValueError("_sub_steps must be positive")

File: src/sc_neurocore/accel/rust/safety/dcn_neuron.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Rust safety for dcn_neuron

#![allow(unused_variables, dead_code, non_snake_case)]

#[derive(Debug, Clone)]
pub struct DCNNeuron {
    pub v: f64,
    pub h: f64,
    pub n: f64,
    pub p: f64,
    pub s: f64,
    pub r: f64,
    pub ca: f64,
    pub g_na: f64,
    pub g_nap: f64,
    pub g_k: f64,
    pub g_t: f64,
    pub g_ahp: f64,
    pub g_h: f64,
    pub g_l: f64,
    pub e_na: f64,
    pub e_k: f64,
    pub e_ca: f64,
    pub e_h: f64,
    pub e_l: f64,
    pub c_m: f64,
    pub phi: f64,
    pub tau_ca: f64,
    pub kd_ahp: f64,
    pub dt: f64,
    pub v_threshold: f64,
    pub gain: f64,
    pub _sub_steps: f64,
}

impl DCNNeuron {
    pub fn new() -> Self {
        Self {
            v: -60.0_f64,
            h: 0.6_f64,
            n: 0.32_f64,
            p: 0.01_f64,
            s: 0.8_f64,
            r: 0.1_f64,
            ca: 0.05_f64,
            g_na: 35.0_f64,
            g_nap: 0.5_f64,
            g_k: 9.0_f64,
            g_t: 0.1_f64,
            g_ahp: 2.0_f64,
            g_h: 0.02_f64,
            g_l: 0.2_f64,
            e_na: 55.0_f64,
            e_k: -90.0_f64,
            e_ca: 120.0_f64,
            e_h: -40.0_f64,
            e_l: -65.0_f64,
            c_m: 1.0_f64,
            phi: 5.0_f64,
            tau_ca: 150.0_f64,
            kd_ahp: 0.5_f64,
            dt: 0.5_f64,
            v_threshold: -20.0_f64,
            gain: 1.0_f64,
            _sub_steps: 20.0_f64,
        }
    }

    pub fn step(&mut self, i_ext: f64) -> i32 {
        if !validate_dcn_neuron(self) || !i_ext.is_finite() {
            return 0;
        }
        let input = self.gain * i_ext;
        let sub_steps = self._sub_steps as usize;
        let sub_dt = self.dt / self._sub_steps;
        let mut fired = 0;
        let (mut v, mut h, mut n, mut p, mut s_gate, mut r, mut ca) =
            (self.v, self.h, self.n, self.p, self.s, self.r, self.ca);
        for _ in 0..sub_steps {
            let alpha_m = safe_rate(0.1, 35.0, v, 10.0, 1.0);
            let beta_m = 4.0 * (-(v + 60.0) / 18.0).exp();
            let m_inf = alpha_m / (alpha_m + beta_m);
            let alpha_h = 0.07 * (-(v + 58.0) / 20.0).exp();
            let beta_h = 1.0 / (1.0 + (-(v + 28.0) / 10.0).exp());
            let alpha_n = safe_rate(0.01, 34.0, v, 10.0, 0.1);
            let beta_n = 0.125 * (-(v + 44.0) / 80.0).exp();
            let p_inf = 1.0 / (1.0 + (-(v + 48.0) / 5.0).exp());
            let tau_p = 5.0 + 15.0 / (1.0 + ((v + 48.0) / 10.0).powi(2)).max(0.01);
            let mt_inf = 1.0 / (1.0 + (-(v + 52.0) / 5.0).exp());
            let s_inf = 1.0 / (1.0 + ((v + 60.0) / 6.5).exp());
            let tau_s = 20.0 + 50.0 / (1.0 + ((v + 65.0) / 10.0).exp());
            let r_inf = 1.0 / (1.0 + ((v + 80.0) / 10.0).exp());
            let tau_r = 100.0 + 200.0 / (1.0 + ((v + 70.0) / 10.0).exp());
            h = exact_hh_gate(h, alpha_h, beta_h, self.phi, sub_dt);
            n = exact_hh_gate(n, alpha_n, beta_n, self.phi, sub_dt);
            p = exact_relax(p, p_inf, tau_p, sub_dt);
            s_gate = exact_relax(s_gate, s_inf, tau_s, sub_dt);
            r = exact_relax(r, r_inf, tau_r, sub_dt);
            let i_t = self.g_t * mt_inf.powi(2) * s_gate * (v - self.e_ca);
            let ca_entry = if i_t < 0.0 { -i_t * 0.001 } else { 0.0 };
            ca = exact_relax(ca, ca_entry * self.tau_ca, self.tau_ca, sub_dt).max(0.0);
            let ahp_inf = ca.powi(2) / (ca.powi(2) + self.kd_ahp.powi(2));
            let g_na_eff = self.g_na * m_inf.powi(3) * h;
            let g_nap_eff = self.g_nap * p;
            let g_k_eff = self.g_k * n.powi(4);
            let g_t_eff = self.g_t * mt_inf.powi(2) * s_gate;
            let g_ahp_eff = self.g_ahp * ahp_inf;
            let g_h_eff = self.g_h * r;
            v = exact_voltage_step(
                v,
                input,
                self.c_m,
                sub_dt,
                &[
                    (g_na_eff, self.e_na),
                    (g_nap_eff, self.e_na),
                    (g_k_eff, self.e_k),
                    (g_t_eff, self.e_ca),
                    (g_ahp_eff, self.e_k),
                    (g_h_eff, self.e_h),
                    (self.g_l, self.e_l),
                ],
            );
            if v >= self.v_threshold {
                fired = 1;
                v = -60.0;
                s_gate *= 0.5;
                ca += 0.5;
            }
        }
        if ![v, h, n, p, s_gate, r, ca]
            .iter()
            .all(|value| value.is_finite())
        {
            return 0;
        }
        self.v = v.clamp(-100.0, 60.0);
        self.h = h.clamp(0.0, 1.0);
        self.n = n.clamp(0.0, 1.0);
        self.p = p.clamp(0.0, 1.0);
        self.s = s_gate.clamp(0.0, 1.0);
        self.r = r.clamp(0.0, 1.0);
        self.ca = ca.max(0.0);
        fired
    }

    pub fn reset(&mut self) {
        self.v = -60.0_f64;
        self.h = 0.6_f64;
        self.n = 0.32_f64;
        self.p = 0.01_f64;
        self.s = 0.8_f64;
        self.r = 0.1_f64;
        self.ca = 0.05_f64;
    }
}

pub fn validate_dcn_neuron(state: &DCNNeuron) -> bool {
    [
        state.v,
        state.h,
        state.n,
        state.p,
        state.s,
        state.r,
        state.ca,
        state.g_na,
        state.g_nap,
        state.g_k,
        state.g_t,
        state.g_ahp,
        state.g_h,
        state.g_l,
        state.e_na,
        state.e_k,
        state.e_ca,
        state.e_h,
        state.e_l,
        state.c_m,
        state.phi,
        state.tau_ca,
        state.kd_ahp,
        state.dt,
        state.v_threshold,
        state.gain,
        state._sub_steps,
    ]
    .iter()
    .all(|value| value.is_finite())
        && [state.h, state.n, state.p, state.s, state.r]
            .iter()
            .all(|gate| (0.0..=1.0).contains(gate))
        && state.ca >= 0.0
        && (-100.0..=60.0).contains(&state.v)
        && [
            state.g_na,
            state.g_nap,
            state.g_k,
            state.g_t,
            state.g_ahp,
            state.g_h,
            state.g_l,
        ]
        .iter()
        .all(|g| *g >= 0.0)
        && state.c_m > 0.0
        && state.phi > 0.0
        && state.tau_ca > 0.0
        && state.kd_ahp > 0.0
        && state.dt > 0.0
        && state.gain >= 0.0
        && state._sub_steps >= 1.0
}

fn safe_rate(a: f64, vhalf: f64, v: f64, k: f64, fallback: f64) -> f64 {
    let d = v + vhalf;
    if d.abs() < 1e-7 {
        fallback
    } else {
        a * d / (1.0 - (-d / k).exp())
    }
}

fn exact_relax(value: f64, target: f64, tau: f64, dt: f64) -> f64 {
    target + (value - target) * (-dt / tau).exp()
}

fn exact_hh_gate(value: f64, alpha: f64, beta: f64, phi: f64, dt: f64) -> f64 {
    let rate = phi * (alpha + beta);
    let target = alpha / (alpha + beta);
    target + (value - target) * (-rate * dt).exp()
}

fn exact_voltage_step(
    v: f64,
    input_current: f64,
    c_m: f64,
    dt: f64,
    conductances: &[(f64, f64)],
) -> f64 {
    let g_total: f64 = conductances.iter().map(|(g, _)| *g).sum();
    if g_total <= 0.0 {
        return v + dt * input_current / c_m;
    }
    let reversal_drive: f64 = conductances.iter().map(|(g, e_rev)| g * e_rev).sum();
    let v_inf = (input_current + reversal_drive) / g_total;
    v_inf + (v - v_inf) * (-dt * g_total / c_m).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dcn_neuron_new() {
        let state = DCNNeuron::new();
        assert!(state.v.is_finite());
        assert!(validate_dcn_neuron(&state));
    }

    #[test]
    fn test_dcn_neuron_step() {
        let mut state = DCNNeuron::new();
        let spike = state.step(10.0);
        assert!(spike == 0 || spike == 1);
        assert!(state.v.is_finite());
        assert!(state.ca >= 0.0);
    }

    #[test]
    fn test_dcn_neuron_closed_form_gate_and_calcium_kinetics() {
        let mut state = DCNNeuron::new();
        state.g_na = 0.0;
        state.g_nap = 0.0;
        state.g_k = 0.0;
        state.g_t = 0.0;
        state.g_ahp = 0.0;
        state.g_h = 0.0;
        state.g_l = 0.0;
        state.gain = 0.0;
        let (v0, h0, n0, p0, s0, r0, ca0) = (
            state.v, state.h, state.n, state.p, state.s, state.r, state.ca,
        );
        let alpha_h = 0.07 * (-(v0 + 58.0) / 20.0).exp();
        let beta_h = 1.0 / (1.0 + (-(v0 + 28.0) / 10.0).exp());
        let alpha_n = safe_rate(0.01, 34.0, v0, 10.0, 0.1);
        let beta_n = 0.125 * (-(v0 + 44.0) / 80.0).exp();
        let p_inf = 1.0 / (1.0 + (-(v0 + 48.0) / 5.0).exp());
        let tau_p = 5.0 + 15.0 / (1.0 + ((v0 + 48.0) / 10.0).powi(2)).max(0.01);
        let s_inf = 1.0 / (1.0 + ((v0 + 60.0) / 6.5).exp());
        let tau_s = 20.0 + 50.0 / (1.0 + ((v0 + 65.0) / 10.0).exp());
        let r_inf = 1.0 / (1.0 + ((v0 + 80.0) / 10.0).exp());
        let tau_r = 100.0 + 200.0 / (1.0 + ((v0 + 70.0) / 10.0).exp());

        state.step(0.0);

        assert_close(state.v, v0);
        assert_close(
            state.h,
            exact_hh_gate(h0, alpha_h, beta_h, state.phi, state.dt),
        );
        assert_close(
            state.n,
            exact_hh_gate(n0, alpha_n, beta_n, state.phi, state.dt),
        );
        assert_close(state.p, exact_relax(p0, p_inf, tau_p, state.dt));
        assert_close(state.s, exact_relax(s0, s_inf, tau_s, state.dt));
        assert_close(state.r, exact_relax(r0, r_inf, tau_r, state.dt));
        assert_close(state.ca, exact_relax(ca0, 0.0, state.tau_ca, state.dt));
    }

    #[test]
    fn test_dcn_neuron_invalid_drive_preserves_state() {
        let mut state = DCNNeuron::new();
        let before = state.clone();
        assert_eq!(state.step(f64::NAN), 0);
        assert_eq!(state.v, before.v);
        assert_eq!(state.ca, before.ca);
    }

    fn assert_close(observed: f64, expected: f64) {
        assert!(
            (observed - expected).abs() <= 1.0e-12,
            "observed {:.17e}, expected {:.17e}",
            observed,
            expected,
        );
    }
}
